File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists == null || lists.length == 0){
            return null;
        }
        return mergeSort(lists,0,lists.length-1);
    }

    private ListNode mergeSort(ListNode[] lists , int start, int end){
        if (start > end){
            return null;
        }
        if (start == end){
            return lists[start];
        }

        int mid = (end+start)/2;
        ListNode left = mergeSort(lists,start, mid);
        ListNode right = mergeSort(lists, mid+1,end);
        return merge(left,right);
    }

    private ListNode merge(ListNode left, ListNode right){
        ListNode result = new ListNode(0);
        ListNode it = result;
        while (left !=null && right !=null){
            if(left.val <= right.val){
                it.next = left;
                left = left.next;
            }else{
                it.next = right;
                right = right.next;
            }
            it = it.next;
        }
        
        if (left !=null){
            it.next = left;
        }else{
            it.next = right;
        }

        return result.next;
    }
}

File: Data Structures & Algorithms/single-element-in-a-sorted-array/submission-0.py
class Solution:
    def singleNonDuplicate(self, nums: List[int]) -> int:
        l,r = 0, len(nums)-1

        while l<r:
            mid = (r+l) //2

            if nums[mid] == nums[mid+1]:
                if (r - (mid+1)) %2 == 0:
                    r = mid-1
                else:
                    l = mid+2
            
            elif nums[mid-1] == nums[mid]:
                if ((mid-1)-l) %2 == 0:
                    l = mid+1
                else:
                    r = mid-2
            else:
                return nums[mid]
        
        return nums[l]